libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/ViewportDataSourceProcessor.kt

```diff
@@ -0,0 +1,328 @@
+     */
+    fun processRouteIntersections(
+        enabled: Boolean,
+        minimumMetersForIntersectionDensity: Double,
+        route: DirectionsRoute,
+        completeRoutePoints: List<List<List<Point>>>
+    ): List<List<Double>> {
+        if (!enabled) {
+            return emptyList()
+        }
+        return route.legs()?.mapIndexed { legIndex, leg ->
+            leg.steps()?.mapIndexed { stepIndex, step ->
+                val stepPoints = completeRoutePoints[legIndex][stepIndex]
+                val intersectionLocations =
+                    step.intersections()?.map { it.location() } ?: emptyList()
+                val list: MutableList<Point> = ArrayList()
+                list.add(stepPoints.first())
+                list.addAll(intersectionLocations)
+                list.add(stepPoints.last())
+                val comparisonList = list.toMutableList()
+                list.removeFirst()
+                val intersectionDistances = list.mapIndexed { index, point ->
+                    TurfMeasurement.distance(point, comparisonList[index]) * 1000.0
+                }
+                val filteredIntersectionDistances =
+                    intersectionDistances.filter { it > minimumMetersForIntersectionDensity }
+                if (filteredIntersectionDistances.isNotEmpty()) {
+                    filteredIntersectionDistances
+                        .reduce { acc, next -> acc + next }
+                        .div(filteredIntersectionDistances.size)
+                } else {
+                    minimumMetersForIntersectionDensity
+                }
+            } ?: emptyList()
+        } ?: emptyList()
+    }
+
+    /**
+     * Returns points to be framed on the remainder of the current step.
+     */
+    fun getPointsToFrameOnCurrentStep(
+        intersectionDensityCalculationEnabled: Boolean,
+        intersectionDensityAverageDistanceMultiplier: Double,
+        averageIntersectionDistancesOnRoute: List<List<Double>>,
+        currentLegProgress: RouteLegProgress,
+        currentStepProgress: RouteStepProgress
+    ): List<Point> {
+        var distanceTraveledOnStepKM =
+            currentStepProgress.distanceTraveled.metersToKilometers().coerceAtLeast(0.0)
+        val fullDistanceOfCurrentStepKM =
+            (currentStepProgress.distanceRemaining + currentStepProgress.distanceTraveled)
+                .metersToKilometers().coerceAtLeast(0.0)
+        if (distanceTraveledOnStepKM > fullDistanceOfCurrentStepKM) {
+            distanceTraveledOnStepKM = 0.0
+        }
+
+        var lookaheadDistanceForZoom = fullDistanceOfCurrentStepKM
+
+        if (intersectionDensityCalculationEnabled &&
+            averageIntersectionDistancesOnRoute.isNotEmpty()
+        ) {
+            val lookaheadInKM = averageIntersectionDistancesOnRoute
+                .get(currentLegProgress.legIndex)
+                .get(currentStepProgress.stepIndex)
+                .div(1000.0)
```

Comment:
Same as above re: magic number (although here it's more evident that it's 1000 meters)